pkg/k8s/apis/cilium.io/v2/client/register.go

```diff
@@ -245,89 +498,102 @@ func createUpdateCRD(clientset apiextensionsclient.Interface,
 		}
 	}
 
-	// wait for the CRD to be established
+	return nil
+}
+
+func waitForV1CRD(
+	scopedLog *logrus.Entry,
+	crdName string,
+	crd *apiextensionsv1.CustomResourceDefinition,
+	client v1client.CustomResourceDefinitionsGetter,
+) error {
 	scopedLog.Debug("Waiting for CRD (CustomResourceDefinition) to be available...")
-	err = wait.Poll(500*time.Millisecond, 60*time.Second, func() (bool, error) {
-		for _, cond := range clusterCRD.Status.Conditions {
+
+	err := wait.Poll(500*time.Millisecond, 60*time.Second, func() (bool, error) {
+		for _, cond := range crd.Status.Conditions {
 			switch cond.Type {
 			case apiextensionsv1.Established:
 				if cond.Status == apiextensionsv1.ConditionTrue {
-					return true, err
+					return true, nil
 				}
 			case apiextensionsv1.NamesAccepted:
 				if cond.Status == apiextensionsv1.ConditionFalse {
-					scopedLog.WithError(goerrors.New(cond.Reason)).Error("Name conflict for CRD")
+					err := goerrors.New(cond.Reason)
+					scopedLog.WithError(err).Error("Name conflict for CRD")
 					return false, err
 				}
 			}
 		}
-		clusterCRD, err = clientset.ApiextensionsV1().CustomResourceDefinitions().Get(
+
+		var err error
+		if crd, err = client.CustomResourceDefinitions().Get(
 			context.TODO(),
 			crd.ObjectMeta.Name,
-			metav1.GetOptions{})
-		if err != nil {
+			metav1.GetOptions{}); err != nil {
 			return false, err
 		}
 		return false, err
 	})
 	if err != nil {
-		deleteErr := clientset.ApiextensionsV1().CustomResourceDefinitions().Delete(
+		if deleteErr := client.CustomResourceDefinitions().Delete(
 			context.TODO(),
 			crd.ObjectMeta.Name,
-			metav1.DeleteOptions{})
-		if deleteErr != nil {
+			metav1.DeleteOptions{},
+		); deleteErr != nil {
 			return fmt.Errorf("unable to delete k8s %s CRD %s. Deleting CRD due: %s",
 				crdName, deleteErr, err)
 		}
 		return err
 	}
```

Comment:
Now I see why you added the Delete permissions. We should remove this code entirely and backported to all branches. If we delete the CRD we will delete all CNPs.

Reply:
Sure, added as a followup in https://github.com/cilium/cilium/issues/12737 so that we can easily backport it. I will drop the commit that adds delete permissions in this PR.